### PR TITLE
implement metadata for metafeed/add/derived, add metadata test

### PR DESCRIPTION
**Pull request by cryptix**
updates #22

### CODE REVIEW

**Comment by cblgh** [comment]
@cryptix lmk what you think about the metadata implementation, thought it was quite clean to introduce the metadata as a  separate method on `AddDerived`. i'm up for changing it if you think that's necessary though!

File: metamngmt/encoding_add_test.go [hunk 1]
```diff
@@ -29,37 +29,41 @@ func TestAddDereivedWithoutMetadata(t *testing.T) {
...
+	metadata := map[string]string{
 		"querylang": "ql-0",
 		"query":     "somejson",
 	}
-	// TODO: add err to New...()
-	addMsg, err := NewAddDerivedMessage(f1, f2, "test", []byte("asdasdasd"), meta)
+	err = addMsg.InsertMetadata(metadata)
+	r.NoError(err)
+
+	// // TODO: update want with actual ouput
```
Comment (cryptix) [inline metamngmt/encoding_add_test.go]:
```suggestion
```

File: metamngmt/msgtypes.go [hunk 2]
```diff
@@ -59,10 +60,13 @@ type AddDerived struct {
...
 	Tangles refs.Tangles `json:"tangles"`
+	QueryLang string `json:"querylang"`
+	Query string `json:"query"`
 }
 
 // NewAddDerivedMessage just initializes type and the passed fields.
-// Callers need to set the right tangle point themselves afterwards.
+// Callers need to set the right tangle point themselves afterwards. If the message has metadata that needs to be added,
+// function AddDerived.InsertMetadata should be used.
 func NewAddDerivedMessage(meta, sub refs.FeedRef, purpose string, nonce []byte) AddDerived {
```
Comment (cryptix) [inline metamngmt/msgtypes.go:70]:
```suggestion
func NewAddDerivedMessage(meta, sub refs.FeedRef, purpose string, nonce []byte, metadata ...map[string]string) (AddDerived, error) {

  if len(metadata) > 0 {

    if len(metadata) > 1 {
      return errors.New("only supports one metadata value")
    }

    err:=insertMetadata(metadata[0])
    // if err...

  }
...
```

This way we don't have to add `InsertMetadata()`.

File: metamngmt/msgtypes.go [hunk 3]
```diff
@@ -59,10 +60,13 @@ type AddDerived struct {
 	Nonce []byte `json:"nonce"`
 
 	Tangles refs.Tangles `json:"tangles"`
+	QueryLang string `json:"querylang"`
+	Query string `json:"query"`
```
Comment (cryptix) [inline metamngmt/msgtypes.go]: Thinking about this some more.. I think it's not good to export them like this.

With this approach can't differentiate the absence of the fields in the bencode data from empty string values.

```suggestion
      metadata map[string]string
```

Plus a function ala `HasMetadata(key string) (string, bool)`

**Review by cryptix — COMMENTED** [review]
Nice, just like I ordered it..!

But I realized there was something wrong with my order.. 🙈

File: metamngmt/encoding_add.go [hunk 4]
```diff
@@ -38,23 +41,32 @@ func (a AddDerived) MarshalBencode() ([]byte, error) {
...
 		Type:        bencodeext.String(typeAddDerived),
-		FeedPurpose: bencodeext.String(a.FeedPurpose),
+		FeedPurpose: bencodeext.String(derived.FeedPurpose),
 		SubFeed:     sfBytes,
 		MetaFeed:    mfBytes,
-		Nonce:       bencodeext.Bytes(a.Nonce),
-		Tangles:     tanglesToBencoded(a.Tangles),
+		Nonce:       bencodeext.Bytes(derived.Nonce),
+		Tangles:     tanglesToBencoded(derived.Tangles),
+		QueryLang: bencodeext.String(querylang),
```
Comment (cryptix) [inline metamngmt/encoding_add.go]: gofmt please :)

**Comment by cblgh** [comment]
hyyeeeeeeeeeeee

**Pull request merged**